### PR TITLE
2581 Joplin content main page

### PR DESCRIPTION
**Pull request by robbobfrh84**
# Description

This will make our "search page" and our "main content page" the same thing for Joplin.
- It will also allow us to filter by `department` and order by `content type` within the search results. 
- See`base/views/joplin_views.py` for much of the core functionality and search handling. 

I've added notes to the code I figured would be valuable for us devs in the future. As well as more details explaining this PR specifically through github comments in the `files changed` tab.

Also, several of the files were "pulled" from wagtails repository and modified for our needs. I tried to keep those files as intact as possible and noted out what I've added/changes. 
- wagtail repo: https://github.com/wagtail/wagtail

### How "Home" and "Root" are handled
- The default pages in search results don't show the "home" and "root" pages.
- However, search for "Root" or "Home" and it'll still show up.

# Fixes issue #2581

* Issue Link: https://app.zenhub.com/workspaces/techstack-5a78b88e1ce69f3510b678ef/issues/cityofaustin/techstack/2581

# Testing Notes

... how can this be tested ?
- Click deployed link👇. 
- Sort a column ("title"/"status"/etc...)
- Search for something...
- Sort that. 
- Filter by department 
- ... does it work?

* Deployed Link🚀:  https://joplin-pr-2581-content-main-pa.herokuapp.com/admin/pages/search/

# Known issues / bugs 

- When ordering by column, you may notice that the url is just getting concatenated. It still returns the correct search but creates awful and long urls. 
  - An issue has been crated for this... https://github.com/cityofaustin/techstack/issues/3261

- Department pages can be sorted, however, it isn't in alphabetical order. I don't know why. I also don't know if it's an issue. 🤷‍♂️. 

# Checklist:
- [x] Request reviewers
- [x] Slack-out message for review
- [x] Link this PR in the issue
- [x] Moved card to *review* in zenhub

### CODE REVIEW

File: joplin/base/views/joplin_search_views.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            if ordering == 'content_type':
+                pages = pages.order_by('content_type')
+            elif ordering == '-content_type':
+                pages = pages.order_by('-content_type')
+
+            # The owner order column was added so Joplin search results could order alphabetically by owner
+            if ordering == 'owner':
+                pages = pages.order_by('owner')
+            elif ordering == '-owner':
+                pages = pages.order_by('-owner')
```
Comment (robbobfrh84) [inline joplin/base/views/joplin_search_views.py]: The `content_type` and `owner` field weren't being ordered when clicking the column header. To add the functionality, their names needed to be added to this condition here.

File: joplin/base/views/joplin_search_views.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            # See: "Content Type Builder"
+
+    else:
+        form = SearchForm()
+        # JOPLIN NOTE: This is where we "hide" the home and root page on initial load of main content page.
+        # - However, these pages will be available in any search that matches title.
+        for page in pages:
+            if page.title == "Root" or page.title == "Home":
+                pages = pages.not_page(page)
+                all_pages = all_pages.not_page(page)
```
Comment (robbobfrh84) [inline joplin/base/views/joplin_search_views.py]: Here is where I filter out the "Root" and "Home" page from the default search view of the main page.

File: joplin/templates/wagtailadmin/pages/search.html [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  </script>
+{% endblock %}
+
+{% block content %}
+  {% trans "Search" as search_str %}
+  {% include "wagtailadmin/shared/header.html" with title=search_str search_url="wagtailadmin_pages:search" icon="search" %}
+
+  <div id="page-results">
+      {% include "wagtailadmin/pages/search_results.html" with query_string=query_string %}
+  </div>
```
Comment (robbobfrh84) [inline joplin/templates/wagtailadmin/pages/search.html]: Here is where can control what variables we can pass from the `joplin_search_views.js` to `search_results.html`
- We needed to pull this up from the wagtail root to modify this transation.

File: joplin/templates/wagtailadmin/pages/search_results.html [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+                  {% endfor %}
+              </ul>
+          </nav>
+        </div>
+
+        <!--
+          This is where Joplin search results are modified for the UI for search results
+          on our search results page.
+        -->
+        {% include "wagtailadmin/pages/listing/_list_explore.html" with show_parent=0 sortable=1 sortable_by_type=0 full_width=1 ordering=request.GET.ordering show_ordering_column=1 orderable=1 %}
```
Comment (robbobfrh84) [inline joplin/templates/wagtailadmin/pages/search_results.html]: Again, this file was brought up from the wagtail root lib so we could modify the headers in the search results UI. Here's where we can also add/remove column headers and behaviors

File: joplin/base/wagtail_hooks.py [hunk 5]
```diff
@@ -72,8 +72,8 @@ def configure_main_menu(request, menu_items):
 
 @hooks.register('register_admin_menu_item')
 def register_page_list_menu_item():
-    home = HomePage.objects.first()
-    return MenuItem('Pages', reverse('wagtailadmin_explore', args=[home.pk]), classnames='icon icon-home', order=10)
+    # home = HomePage.objects.first()
```
Comment (briaguya) [inline joplin/base/wagtail_hooks.py]: so glad we don't have the hacky /3/ logic anymore

File: joplin/base/views/joplin_search_views.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
+from django.shortcuts import render
+from django.http.request import QueryDict
+from django.contrib.contenttypes.models import ContentType
+from django.core.paginator import Paginator
+from wagtail.core.models import Page
+from wagtail.search.query import MATCH_ALL
+from wagtail.admin.forms.search import SearchForm
+
+def search(request):
```
Comment (easherma) [inline joplin/base/views/joplin_search_views.py]: Lets add a docstring here explaining that this is for an why it exists

File: joplin/base/views/joplin_search_views.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        except ContentType.DoesNotExist:
+            raise Http404
+
+        pages = pages.filter(content_type=selected_content_type)
+    else:
+        selected_content_type = None
+
+    query = ''
+
+    # JOPLIN NOTE: Some of this the original state of the query condition
```
Comment (easherma) [inline joplin/base/views/joplin_search_views.py]: multiline comments are much easier to read/write as docstrings

File: joplin/templates/wagtailadmin/pages/index.html [hunk 8]
```diff
@@ -1,26 +0,0 @@
-{% extends "wagtailadmin/pages/index.html" %}
```
Comment (easherma) [inline joplin/templates/wagtailadmin/pages/index.html]: Why is this deleted?

File: joplin/templates/wagtailadmin/pages/listing/_table_headers_explore.html [hunk 9]
```diff
@@ -20,16 +20,16 @@
         <th class="ord">
```
Comment (easherma) [inline joplin/templates/wagtailadmin/pages/listing/_table_headers_explore.html]: I see what you're doing here and it makes sense. However, rather than edit this table_headers_explore file, I think we should edit/extend the template used by wagtail for search. We're now turning this template into something that it isn't, and it isn't going to be clear in the future why we did that and may cause future compatibility problems or make it difficult to switch back. 

At first glance it looks like search is using this template, perhaps? 
https://github.com/wagtail/wagtail/blob/23a89774adef4f609e33eb3feece90d6535588ee/wagtail/admin/templates/wagtailadmin/pages/search_results.html

File: joplin/urls.py [hunk 10]
```diff
@@ -9,14 +9,16 @@
 from wagtail.core import urls as wagtail_urls
 from wagtail.documents import urls as wagtaildocs_urls
 from base.views import joplin_views
+from base.views import joplin_search_views
 from django.urls import reverse
-from base.models import HomePage
+# from base.models import HomePage
```
Comment (easherma) [inline joplin/urls.py]: lets remove stuff rather than commenting it out. 

However, I'm wondering if we might still be able to leverage the HomePage model for filtering out on the search view....

File: joplin/base/views/joplin_search_views.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+            if ordering == 'latest_revision_created_at':
+                pages = pages.order_by('latest_revision_created_at')
+            elif ordering == '-latest_revision_created_at':
+                pages = pages.order_by('-latest_revision_created_at')
+
+            if ordering == 'live':
+                pages = pages.order_by('live')
+            elif ordering == '-live':
+                pages = pages.order_by('-live')
```
[the remark this answers is not in the document]
Comment (robbobfrh84) [reply joplin/base/views/joplin_search_views.py]: Good point. I'm gonna plug that in and give it a go! Should remove a ton of this code here.

File: joplin/urls.py (hunk 10 above)
Comment (robbobfrh84) [reply joplin/urls.py]: That was kinda my thinking toward "not" moving it out. I wanted to see y'all perspective on if you wanted it or not. 

The thoughts I had about this were, Our search page is our home page. But, that pages is also a search page... by it's very nature. 

I'm gonna take it out and put that note in.

File: joplin/templates/wagtailadmin/pages/listing/_table_headers_explore.html (hunk 9 above)
Comment (robbobfrh84) [reply joplin/templates/wagtailadmin/pages/listing/_table_headers_explore.html]: So that is the wagtail template file that I did pull in here: 
- ttps://github.com/cityofaustin/joplin/pull/418/files?file-filters%5B%5D=.html&file-filters%5B%5D=.py&hide-deleted-files=true#diff-8d43fac787c5b9a5e1af51737d716906

In addition the `_table_headers_explore.html` file was also pulled in from wagtails templates here: 
- https://github.com/wagtail/wagtail/blob/23a89774adef4f609e33eb3feece90d6535588ee/wagtail/admin/templates/wagtailadmin/pages/listing/_table_headers_explore.html
- Then, I only modified the url when sorting to *included* the search query so it didn't revert back to *all* pages when sorting.

Comment (robbobfrh84) [reply joplin/templates/wagtailadmin/pages/listing/_table_headers_explore.html]: I'll add notes to that file being pulled in so it's clear where it came from.

File: joplin/templates/wagtailadmin/pages/index.html (hunk 8 above)
Comment (robbobfrh84) [reply joplin/templates/wagtailadmin/pages/index.html]: The functionality of this file was added to the new `search_header.html`. It was showing up double or in unwanted places.

File: joplin/base/views/joplin_search_views.py (hunk 7 above)
Comment (robbobfrh84) [reply joplin/base/views/joplin_search_views.py]: I've added those in the updated PR

File: joplin/base/wagtail_hooks.py (hunk 5 above)
Comment (easherma) [reply joplin/base/wagtail_hooks.py]: Thats been changed for awhile

File: joplin/base/views/joplin_search_views.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (easherma) [reply joplin/base/views/joplin_search_views.py]: This might be a good contrib back to wagtail core!
